### ISSUE TITLE
[FR] Experiment Tracking Decorator to remove boilerplate

### ISSUE THREAD

**Issue by Chris-hughes10** [mlflow/mlflow#3950]
Thank you for submitting a feature request. **Before proceeding, please review MLflow's [Issue Policy for feature requests](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md#feature-requests) and the [MLflow Contributing Guide](https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.rst)**.

**Please fill in this feature request template to ensure a timely and thorough response.**

## Willingness to contribute
The MLflow Community encourages new feature contributions. Would you or another member of your organization be willing to contribute an implementation of this feature (either as an MLflow Plugin or an enhancement to the MLflow code base)?

- [ ] Yes. I can contribute this feature independently.
- [x] Yes. I would be willing to contribute this feature with guidance from the MLflow community.
- [ ] No. I cannot contribute this feature at this time.

## Proposal Summary

With the current tracking API, some boilerplate code is required to create an experiment and start a run. Specifically, it is quite common to see code blocks such as:
```
mlflow.set_tracking_uri(MLFLOW_TRACKING_SERVICE_URI)
experiment_id = mlflow.set_experiment(EXPERIMENT_NAME)
with mlflow.start_run(experiment_id=experiment_id):
    ...
``` 
at the start of a training script.

In an effort to remove this when using mlflow for personal projects, I implemented a decorator that can be added to a training function, which can set some of these parameters using decorator arguments or environment variables. The syntax currently looks like:
```
@mlflow_experiment
def train_model_1():
   ...

@mlflow_experiment(experiment_name='Train Random Forest', autolog=True)
def train_sklearn_model():
    ...

This has worked well for me and has resulted in my training scripts requiring less mlflow specific code - especially when combined with autologging - and eases adoption. I have primarily used this in scripts that are deployed on platforms such as AzureML, Kubeflow and Argo as opposed to in a notebook environment. This approach has received positive feedback from my colleagues at Microsoft, and I believe this could be a useful feature for other members of the community, and as a single function which uses existing components, it would be simple to integrate.

## Motivation
- What is the use case for this feature?
To remove boilerplate code from training scripts when using the tracking API for simple use cases.
- Why is this use case valuable to support for MLflow users in general?
Less code is required to start using mlflow, and it arguably provides a cleaner approach than a context manager.
- Why is this use case valuable to support for your project(s) or organization?
This enabled me to remove some boilerplate code that I was observing in every training script in which I was using mlflow
- Why is it currently difficult to achieve this use case? (please be as specific as possible about why related MLflow features and components are insufficient)
Currently, the use, and knowledge, of additional components in the tracking API are required to create experiments and log results to a remote tracking server, which ends up becoming boilerplate.

### What component(s), interfaces, languages, and integrations does this feature affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [x] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interfaces
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Languages 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

## Details

(Use this section to include any additional information about the feature. If you have a proposal for how to implement this feature, please include it here. For implementation guidelines, please refer to the [Contributing Guide](https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.rst#contribution-guidelines).)

I shall include the implementation (and tests) that I am currently using, which have worked well for my use cases, below:
```
import functools
import os

import mlflow

def mlflow_experiment(
    _func=None,
    *,
    experiment_name=None,
    tracking_uri=None,
    autolog=False,
    run_name=None,
    tags=None,
):
    def experiment_decorator(func):
        @functools.wraps(func)
        def experiment_wrapper(*args, **kwargs):
            nonlocal experiment_name, tracking_uri

            if tracking_uri is None:
                tracking_uri = os.getenv(
                    "MLFLOW_TRACKING_SERVICE_URI", mlflow.get_tracking_uri()
                )
            mlflow.set_tracking_uri(tracking_uri)

            if experiment_name is None:
                experiment_name = os.getenv("MLFLOW_EXPERIMENT_NAME", None)

            experiment_id = (
                mlflow.set_experiment(experiment_name)
                if experiment_name is not None
                else None
            )

            if autolog:
                mlflow.autolog()

            with mlflow.start_run(
                experiment_id=experiment_id, run_name=run_name, tags=tags
            ):
                value = func(*args, **kwargs)

            return value

        return experiment_wrapper

    if _func is None:
        return experiment_decorator
    else:
        return experiment_decorator(_func)

---

from unittest.mock import MagicMock

from pytest import fixture

from mlflow_decorator import mlflow_experiment

@fixture
def target_func():
    return MagicMock()

EXPERIMENT_NAME = "experiment_name"
EXPERIMENT_ID = "experiment_id"
TRACKING_URI = "tracking_uri"
RUN_NAME = "run_name"
TAGS = "tags"
FUNCTION_ARG = "arg1"
FUNCTION_KWARG = "kwarg1"

MLFLOW_TRACKING_SERVICE_URI_ENV_VAR = "MLFLOW_TRACKING_SERVICE_URI_ENV_VAR"
MLFLOW_EXPERIMENT_NAME_ENV_VAR = "MLFLOW_EXPERIMENT_NAME_ENV_VAR"

def get_env_fake(env_name, *args):
    if env_name == "MLFLOW_TRACKING_SERVICE_URI":
        return MLFLOW_TRACKING_SERVICE_URI_ENV_VAR
    elif env_name == "MLFLOW_EXPERIMENT_NAME":
        return MLFLOW_EXPERIMENT_NAME_ENV_VAR

def test_experiment_decorator_no_args(mocker, target_func):
    mlflow_mock = MagicMock()
    get_tracking_uri_mock = MagicMock(return_value=TRACKING_URI)
    mlflow_mock.get_tracking_uri = get_tracking_uri_mock
    mocker.patch("mlflow_decorator.mlflow", mlflow_mock)

    decorated_func = mlflow_experiment(target_func)
    decorated_func(FUNCTION_ARG, FUNCTION_KWARG=FUNCTION_KWARG)

    mlflow_mock.get_tracking_uri.assert_called_once()
    mlflow_mock.set_tracking_uri.assert_called_once_with(TRACKING_URI)
    mlflow_mock.start_run.assert_called_once_with(
        experiment_id=None, run_name=None, tags=None
    )
    target_func.assert_called_once_with(FUNCTION_ARG, FUNCTION_KWARG=FUNCTION_KWARG)

def test_experiment_decorator_with_env_vars(mocker, target_func):
    # patch os.getenv to return specified values
    os_mock = MagicMock()
    os_mock.getenv = get_env_fake
    mocker.patch("mlflow_decorator.os", os_mock)

    mlflow_mock = MagicMock()
    mlflow_mock.set_experiment.return_value = EXPERIMENT_ID
    mocker.patch("mlflow_decorator.mlflow", mlflow_mock)

    decorated_func = mlflow_experiment(target_func)
    decorated_func(FUNCTION_ARG, FUNCTION_KWARG=FUNCTION_KWARG)

    mlflow_mock.set_tracking_uri.assert_called_once_with(
        MLFLOW_TRACKING_SERVICE_URI_ENV_VAR
    )
    mlflow_mock.set_experiment.assert_called_once_with(MLFLOW_EXPERIMENT_NAME_ENV_VAR)
    mlflow_mock.start_run.assert_called_once_with(
        experiment_id=EXPERIMENT_ID, run_name=None, tags=None
    )
    target_func.assert_called_once_with(FUNCTION_ARG, FUNCTION_KWARG=FUNCTION_KWARG)

def test_experiment_decorator_with_args(mocker, target_func):
    mlflow_mock = MagicMock()
    mlflow_mock.set_experiment.return_value = EXPERIMENT_ID
    mocker.patch("mlflow_decorator.mlflow", mlflow_mock)

    decorator_with_args = mlflow_experiment(
        experiment_name=EXPERIMENT_NAME,
        run_name=RUN_NAME,
        tracking_uri=TRACKING_URI,
        tags=TAGS,
    )
    decorated_func = decorator_with_args(target_func)
    decorated_func(FUNCTION_ARG, FUNCTION_KWARG=FUNCTION_KWARG)

    mlflow_mock.set_tracking_uri.assert_called_once_with(TRACKING_URI)
    mlflow_mock.set_experiment.assert_called_once_with(EXPERIMENT_NAME)
    mlflow_mock.start_run.assert_called_once_with(
        experiment_id=EXPERIMENT_ID, run_name=RUN_NAME, tags=TAGS
    )
    target_func.assert_called_once_with(FUNCTION_ARG, FUNCTION_KWARG=FUNCTION_KWARG)

def test_experiment_decorator_can_enable_autologging(mocker, target_func):
    mlflow_mock = MagicMock()
    mocker.patch("mlflow_decorator.mlflow", mlflow_mock)

    decorator_with_args = mlflow_experiment(autolog=True)
    decorated_func = decorator_with_args(target_func)
    decorated_func()

    mlflow_mock.autolog.assert_called_once()
```

**Comment by dmatrix** [comment]
@Chris-hughes10 Thanks for filling this FR. Removing boilerplate code by providing built-in decorators or utility methods is not a bad idea. At the same time, we have to be also careful that we don't introduce or nudge a specfici pattern of writing a training code function to employ a decorator. 

IMHO, while it certainly works well with your use-case of writing a decorated training_func, it may not apply as a general use case. However, I think this might be a worthwhile effort to provide as a PyPI package or extension that people can `pip install mflow[package]` that would enable them to use this decorator pattern for training loops. And perhaps write a blog that shows how to use this extension to write training code in scenarios you have found them useful.

Again, this is only my humble opinion. I will have the MLflow committers opine on this, as this affects in a way how to use or extend MLflow APIs if it's integrated. But having an installable extension offers developers an option if in all their use cases this pattern holds.

cc: @harupy @smurching @dbczumar Any opinion on this?

**Comment by Chris-hughes10** [comment]
Hi @dmatrix, thanks for sharing your thoughts on this! As a last resort, I think a PyPI package would work, but may be overkill for essentially a single function. Additionally, I would be happy to write a blog post or tutorial for mlflow.org. 

With regards to use cases, I think it is pretty general, in my experience it is quite common to see training loops in functions (outside of notebooks) even if this is just a main function. As for the decorator nudging people towards the use of training functions, I have often seen, and advocate, the use of main functions in scripts - as opposed to writing the code directly in `if __name__ == '__main__'` as this makes code easier to test, so it could arguably proposed that such encouragement is a good thing. However, if someone strongly disagrees with this, the existing functionality is still there.

To illustrate a very general use case, here is an adaptation of the tutorial found at https://www.mlflow.org/docs/latest/tutorials-and-examples/tutorial.html.

```
# The data set used in this example is from http://archive.ics.uci.edu/ml/datasets/Wine+Quality
# P. Cortez, A. Cerdeira, F. Almeida, T. Matos and J. Reis.
# Modeling wine preferences by data mining from physicochemical properties. In Decision Support Systems, Elsevier, 47(4):547-553, 2009.
import argparse
import logging
import warnings

import mlflow
import mlflow.sklearn
import numpy as np
import pandas as pd
from sklearn.linear_model import ElasticNet
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.model_selection import train_test_split

from mlflow_decorator import mlflow_experiment

logging.basicConfig(level=logging.WARN)
logger = logging.getLogger(__name__)


def eval_metrics(actual, pred):
    rmse = np.sqrt(mean_squared_error(actual, pred))
    mae = mean_absolute_error(actual, pred)
    r2 = r2_score(actual, pred)
    return rmse, mae, r2


@mlflow_experiment(autolog=True)
def main(alpha, l1_ratio):
    warnings.filterwarnings("ignore")
    np.random.seed(40)

    # Read the wine-quality csv file from the URL
    csv_url = "http://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-red.csv"

    # manual logging still works the same way with autolog enabled
    mlflow.log_param("csv_url", csv_url)

    try:
        data = pd.read_csv(csv_url, sep=";")
    except Exception as e:
        logger.exception(
            "Unable to download training & test CSV, check your internet connection. Error: %s",
            e,
        )
        raise e

    # Split the data into training and test sets. (0.75, 0.25) split.
    train, test = train_test_split(data)

    # The predicted column is "quality" which is a scalar from [3, 9]
    train_x = train.drop(["quality"], axis=1)
    test_x = test.drop(["quality"], axis=1)
    train_y = train[["quality"]]
    test_y = test[["quality"]]

    lr = ElasticNet(alpha=alpha, l1_ratio=l1_ratio, random_state=42)
    lr.fit(train_x, train_y)

    predicted_qualities = lr.predict(test_x)

    (rmse, mae, r2) = eval_metrics(test_y, predicted_qualities)

    print("Elasticnet model (alpha=%f, l1_ratio=%f):" % (alpha, l1_ratio))
    print("  RMSE: %s" % rmse)
    print("  MAE: %s" % mae)
    print("  R2: %s" % r2)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--alpha", default="0.5", required=False)
    parser.add_argument("--l1_ratio", default="0.5", required=False)

    args = parser.parse_args()
    main(args.alpha, args.l1_ratio)
```

As we can see, combined with the use of autologging (which is amazing by the way!) there is hardly any mlflow specific code, which I find remarkable considering what is logged to the dashboard! In fact, in this example, refactoring the training loop to a function which has injected arguments also enables us to remove the confusing lines:
`
alpha = float(sys.argv[1]) if len(sys.argv) > 1 else 0.5
l1_ratio = float(sys.argv[2]) if len(sys.argv) > 2 else 0.5
`
as well as making it easier to test using a tool like pytest. In this case, if we wished to leave the manual logging as in the tutorial, the decorator would simply be `@mlflow_experiment`. 

Hopefully this has helped to illustrate the generality of this approach and the benefits of integrating it into the core package. I look forward to hearing the opinions from other team members and am happy to help where I can.

**Comment by dbczumar** [comment]
@dmatrix @Chris-hughes10 I'm a fan of the core premise here, which seems to be providing a decorator equivalent for `mlflow.start_run()`. I don't see a fundamental reason why an `@mlflow_run` decorator (which I think describes the use case a bit more accurately than `@mlflow_experiment`) couldn't be introduced to MLflow; in fact, it would be even better if `mlflow.start_run()` could be made usable as a decorator directly without introducing a new method (seems tricky). If a new method is indeed necessary, I think the `mlflow.tracking` module would be an appropriate place to house it (e.g., `mlflow.tracking.mlflow_run`).

I have a couple of comments on the parameters:

- It would be nice to match the function signature of `start_run` for consistency (i.e. `experiment_name` should be revised to `experiment_id`
- `set_tracking_uri` operates at a global level, resulting in side-effects beyond the scope of execution of the decorated function. Accordingly, I don't think this is quite the right place for a `tracking_uri` parameter, especially since most use cases will leverage the existing tracking URI without needing to override it
- Similarly, `mlflow.autolog` operates at a global level, enabling autologging for all subsequent runs beyond those created within scope of the decorated function. While I think it's a really interesting idea to consider run-specific autologging enablement in the context of `start_run`, I don't think we should add an `autolog` parameter to such a decorator unless / until we decide to incorporate it into `start_run` as well.

Let me know if this feedback makes sense. From my perspective, I would be more than happy to review a PR with these changes applied. cc @smurching @harupy for further feedback.

**Comment by Chris-hughes10** [comment]
Hi @dbczumar, thanks for your feedback, I'm glad that you like the approach and I'm happy to take your feedback onboard and submit a PR. Looking over the codebase, I think the best option would be introducing an additional `@mlflow_run` function as opposed to modifying `start_run`.

The reasoning behind the current choice of parameters was to enable mlflow to be enabled in an existing script with minimal effort, just by adding a single line; even without really knowing the details of the mlflow API. As the experiment name and tracking uri can be set using environment variables I am happy to remove them as decorator parameters, although this would then require the user to take an additional step to set these. Perhaps just having more visibility on this for new users would suffice in the form of a tutorial or similar alongside the use of the decorator, which I would be happy to contribute as part of this work.

However, with regards to the autologging, I have a couple of suggestions. If it was to be included in the decorator, it could be reverted to its initial state after the function execution has finished, which would remove the side effects and still maintain the benefits of converting an existing script to use mlflow in a single line. Alternatively, although this would be a different feature, what would your thoughts be on using an environment variable to enable autologging globally, which would also somewhat solve the problem?

Once I have clarity on this additional point, I should be able to turn a PR around for this quite quickly as the code is essentially already written.

**Comment by Chris-hughes10** [comment]
CC: @dbczumar I have created a PR based on your feedback: https://github.com/mlflow/mlflow/pull/4065

**Comment by Ilykuleshov** [comment]
Any news on this? This should be stupidly simple to implement with [contextlib context managers](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager), removing the need for ActiveRun:
```
@contextmanager
def start_run(...):
    ... # All code before return stays the same

    # Instead of return we yield
    try:
        yield active_run_stack[-1]
    finally:
        ... # ActiveRun.__exit__ code
```

In the try block, it may be possible to handle specific exceptions and translate them into different finished states, e.g. KeyboardInterrupt -> Killed, other exceptions -> failed, etc.

**Comment by Chris-hughes10** [comment]
@Ilykuleshov I created a PR for this years ago, but it was never reviewed.
